deepcell/training/model_trainer.py

```diff
@@ -0,0 +1,486 @@
+            self.training_callbacks = get_callbacks(
+                model_name, lr_sched=self.lr_sched,
+                tensorboard_log_dir=self.log_dir,
+                save_weights_only=False,
+                monitor='val_loss', verbose=1)
+        else:
+            self.training_callbacks = training_callbacks
+
+        self.trained = False
+
+        self._init_output_metadata()
+
+    def _set_random_seeds(self):
+        pass
+        #os.environ['PYTHONHASHSEED']=str(seed_value)
+        #random.seed(seed_value)
+        #np.random.seed(self.random_seed)
+        #tf.set_random_seed(self.random_seed)
+
+    def _init_output_metadata(self):
+        """
+        self.output_metadata should ultimately contain human-readable and hashed information
+        on the following components of the training process:
+            deepcell version
+            dataset
+            preprocessing function(s)
+            image generators
+            model (along with initialization state?)
+            postprocessing function(s)
+        """
+        # initialize metadata
+        self.output_metadata = {}
+        # store deepcell version
+        self.output_metadata["deepcell_version"] = __version__
+        # store dataset information
+        self.output_metadata["dataset"] = {}
+        self.output_metadata["dataset"]["training"] = {}
+        self.output_metadata["dataset"]["training"]["X_train_shape"] = self.X_train.shape
+        self.output_metadata["dataset"]["training"]["X_train_datatype"] = self.X_train.dtype.name
+        self.output_metadata["dataset"]["training"]["X_train_md5_digest"] = hashlib.md5(self.X_train).hexdigest()
+        self.output_metadata["dataset"]["training"]["y_train_shape"] = self.y_train.shape
+        self.output_metadata["dataset"]["training"]["y_train_datatype"] = self.y_train.dtype.name
+        self.output_metadata["dataset"]["training"]["y_train_md5_digest"] = hashlib.md5(self.y_train).hexdigest()
+        self.output_metadata["dataset"]["testing"] = {}
+        self.output_metadata["dataset"]["testing"]["X_test_shape"] = self.X_test.shape
+        self.output_metadata["dataset"]["testing"]["X_test_datatype"] = self.X_test.dtype.name
+        self.output_metadata["dataset"]["testing"]["X_test_md5_digest"] = hashlib.md5(self.X_test).hexdigest()
+        self.output_metadata["dataset"]["testing"]["y_test_shape"] = self.y_test.shape
+        self.output_metadata["dataset"]["testing"]["y_test_datatype"] = self.y_test.dtype.name
+        self.output_metadata["dataset"]["testing"]["y_test_md5_digest"] = hashlib.md5(self.y_test).hexdigest()
+
+    def _data_prep(self):
+        
+        def prep_generator(
+                generator_type,
+                output_metadata,
+                data_generator,
+                input_data_dict,
+                skip = None,
+                seed = None,
+                batch_size = None,
+                transform = None,
+                transform_kwargs = {}):
```

Comment:
Honestly, I think this would be a bit simpler if `_prep_generator` was just a method instead of an internal function like this.

That way you may not need to pass so many arguments and it can directly rely on the attributes. I think it would be a bit easier to understand.